mersenne-31/src/lib.rs

```diff
@@ -247,14 +247,20 @@ impl Add for Mersenne31 {
     type Output = Self;
 
     fn add(self, rhs: Self) -> Self {
-        let mut sum = self.value + rhs.value;
-        // If sum's most significant bit is set, we clear it and add 1, since 2^31 = 1 mod p.
-        // This addition of 1 cannot overflow 2^31, since sum has a max of
-        // 2 * (2^31 - 1) = 2^32 - 2.
-        let msb = sum & (1 << 31);
-        sum.bitxor_assign(msb);
-        sum += u32::from(msb != 0);
-        Self::new(sum)
+        // Working with i32 means we get a flag which informs us if overflow happened.
+        let (mut sum, over) = (self.value as i32).overflowing_add_unsigned(rhs.value);
+
+        // The base sum will never exceed 2**32 - 1 so as bits sum is exactly the same as self.value + rhs.value
+        // Hence if we didn't overflow we have the correct value.
+        // Otherwise we can simply remove the most significant bit and add 1.
+        // Removing the most significant bit does nothing if no overflow happened.
+        sum.bitand_assign(Self::ORDER_U32 as i32);
+
+        if over {
+            Self::new((sum + 1) as u32)
+        } else {
+            Self::new(sum as u32)
+        }
```

Comment:
You can avoid the condition with:
```suggestion
        Self::new((sum + over as i32) as u32)
```
and similarly for `sub()` below. Compiler might do that transformation automatically, but worth checking whether this is faster.

Reply:
Turns out the compiler treats these two expressions as equivalent, so it's a matter of taste which one to use (up to you, but I'm inclined towards the shorter one). OTOH, at first glance the compiler output for this replacement add does not look better than the previous implementation: https://rust.godbolt.org/z/js4vcEbn6

Reply:
Hmm, very happy to switch to the shorter expression without the if statement. That does feel a little cleaner.

Can also leave the add as is and just change the sub operation. I seemed to be getting a small speed-up for this new add but that could have just been noise and/or the windows compiler behaving strangely.

Reply:
Ahh I had a benchmark in a previous commit but took it out as I wasn't 100% sure on the protocol and wanted to avoid a proliferation of benchmarking files. I've added it back in (Just the one for sub but with a 1 character change it's also a benchmark for add). I'll have a read of how things were done in Plonky2 to see if I can clean it up.

Strangely, running it today I'm also finding that the old add seems to be faster so maybe we just stick with that one. Though the new sub seems to be faster than either option for add.

Reply:
Okay, based of the Plonky2 implementation, I've combined all the current benchmarking files into 1 called bench_field.

Also, playing around with the compiler, I think the function 'overflowing_add_unsigned' was the culprit for why the output assembly was so long. I'd assumed that .overflowing_add_unsigned(rhs) and .overflowing_add(rhs as i32) would act essentially the same but it seems the second option is vastly preferable and produces much less assembly.